### PR TITLE
CQM serialization 2.0

### PR DESCRIPTION
**Pull request by arcondello**
Closes https://github.com/dwavesystems/dimod/issues/1303
Closes https://github.com/dwavesystems/dimod/issues/1302 by creating a new "quadratic" section that encodes the quadratic interactions in COO-format. This will be about 50% larger than necessary in a fully connected quadratic model, but much smaller when the model is linear or sparse. Given that the majority of constraints are sparse, it makes sense to optimize for that case.

### CODE REVIEW

**Comment by arcondello** [comment]
Closing for now. Will re-open after another pass.

File: dimod/constrained/constrained.py [hunk 1]
```diff
@@ -1021,32 +1092,40 @@ def from_file(cls,
 
         header_info = read_header(file_like, CQM_MAGIC_PREFIX)
 
-        if not (1, 0) <= header_info.version <= (1, 3):
+        num_variables = header_info.data["num_variables"]
+
+        if not (1, 0) <= header_info.version <= (2, 0):
```
Comment (randomir) [inline dimod/constrained/constrained.py:1097]: Is `1.0` actually supported?

File: dimod/constrained/constrained.py [hunk 2]
```diff
@@ -1623,41 +1713,55 @@ def to_file(self, *,
...
+                        [penalty]
+                        [weight]
+                    ...
+                objective
+                varinfo
+                [variable_labels.json]
+
+            The ``objective`` file encodes the objective.
+            See :meth:`~dimod.constrained.expression.ObjectiveView._to_file()`
+            for details about the file format.
```
Comment (randomir) [inline dimod/constrained/constrained.py:1745]: It would be nice to have the complete format spec decoupled from dimod internals.

File: dimod/constrained/constrained.py [hunk 3]
```diff
@@ -1771,7 +1851,7 @@ def violations(self, sample_like: SamplesLike, *,
             requirement of a second, and violates a third.
 
             >>> i, j, k = dimod.Binaries(['i', 'j', 'k'])
-            >>> cqm = dimod.ConstrainedQuadraticModel()
+                >>> cqm = dimod.ConstrainedQuadraticModel()
```
Comment (randomir) [inline dimod/constrained/constrained.py:1854]: typo

**Review by randomir — APPROVED** [review]
Looks good to me overall, but I would suggest extending/introducing a bit more of LBYL safety checks (e.g. edge cases when input data is invalid/unexpected/malicious). Also, extending tests beyond these nice high-level ones to catch lower-level decoding/encoding errors of unexpected inputs.

File: dimod/constrained/constrained.py (hunk 1 above)
Comment (arcondello) [reply dimod/constrained/constrained.py:1097]: Yep, the difference is actually quite small. Agree this need a documentation pass to help make this stuff clearer.

File: dimod/constrained/constrained.py (hunk 2 above)
Comment (arcondello) [reply dimod/constrained/constrained.py:1745]: This is just deferring the documentation to another function so that it's easier to maintain. Though overall the documentation for this could/should be improved.

File: dimod/constrained/constrained.py (hunk 3 above)
Comment (arcondello) [inline dimod/constrained/constrained.py:1854]:
```suggestion
            >>> cqm = dimod.ConstrainedQuadraticModel()
```
